src/backend/distributed/transaction/backend_data.c

```diff
@@ -701,19 +704,27 @@ GetCurrentDistributedTransactionId(void)
  * sets it for the current backend. It also sets the databaseId and
  * processId fields.
  *
- * This function should only be called on BeginCoordinatedTransaction(). Any other
- * callers is very likely to break the distributed transction management.
+ * This function returns the same transaction id for consecutive calls from a same
+ * backend. So, it is idempotent.
  */
 void
 AssignDistributedTransactionId(void)
 {
 	pg_atomic_uint64 *transactionNumberSequence =
 		&backendManagementShmemData->nextTransactionNumber;
 
-	uint64 nextTransactionNumber = pg_atomic_fetch_add_u64(transactionNumberSequence, 1);
+	uint64 nextTransactionNumber = 0;
 	int localGroupId = GetLocalGroupId();
 	TimestampTz currentTimestamp = GetCurrentTimestamp();
 
+	/* return idempotent transaction id for consecutive requests from a same session */
+	if (distributedTransactionIdAlreadySet)
+	{
+		return;
+	}
+
+	nextTransactionNumber = pg_atomic_fetch_add_u64(transactionNumberSequence, 1);
```

Comment:
Given that we'll call this function a lot more frequently, we're increasing the probability of overflows. Do you see any issues in case the `transactionNumber` overflows?